### DISCUSSION TITLE
Folder and batch download

### DISCUSSION THREAD

**Discussion by myl7** [spencerwooo/onedrive-vercel-index#136] [Ideas]
onedrive-vercel-index has preview support for many file formats, which encourages webmasters to serve the shared files in bulk other than in zip, making preview of the files possible.
However, onedrive API doesnot provide a way to download a folder or download files in batches. You actually *can* do them via onedrive share url, and onedrive just doesnot provide the api.
The fact makes visitors difficult to download the files stored (even hierarchically) in a folder.
To solve it and enhance the experience, I have come up with some solutions:

1. ... just paste an onedrive share url in the index page. Users can go to the onedrive share page to download anything they want. No code change required.
2. Provide a button to let the user download (even recursively) all items in the page. Seems not hard.
3. Provide a button to copy all download urls in the page, and ask the user to use other tools to download these files. Even easier than 2.
4. Provide a button to send all download urls to a paticular server, and the server fetches the files, zips them, and sends the zip files back to the user in stream. This additionally requires another backend project and makes onedrive-vercel-index not so serverless.
5.  Fetch the files, zip them, and download the zip files in stream, but in client side, in browser js. I am not sure if this is possible. But this will be the most comfortable solution for users.

Currently I would support 1 & 2.

**Comment by opoet** [comment]
是否能像MEGA的下载方式一样，下载到浏览器然后打包取回本地

不知道这个是否有帮助 https://www.zhangxinxu.com/wordpress/2019/08/js-zip-download

**Reply by myl7** [reply]
那这敢情好，就算有内存限制没法给很大的文件夹打包下载，也至少可以用在一些小而文件多的文件夹上

**Comment by myl7** [comment]
After reading the code, I found that the difficulty should be that how to get the recusive file download urls.
The program uses Next API routes to request OneDrive API, and the rendered page uses SWR to cache (and paginate) Next API routes responses.

To reuse SWR to get the recusive download urls, the useSWR React hook needs to be called multiple times in a loop with condition check, which violates React hook rules.
To reuse `pages/api/index.ts` and add extra query params to the handler, or create a new file inside `pages/api` to provide a new API, not only can not reuse the SWR cache, but maybe require the authorization of @spencerwooo
